src/typechecker/FStar.TypeChecker.TcTerm.fs

```diff
@@ -776,24 +776,35 @@ and tc_maybe_toplevel_term env (e:term) : term                  (* type-checked
 
     let guard_x = S.new_bv (Some e1.pos) c1.res_typ in
     let t_eqns = eqns |> List.map (tc_eqn guard_x env_branches) in
-    let c_branches, g_branches =
-      let cases, g =
+    let c_branches, g_branches, erasable =
+      let cases, g, erasable =
           List.fold_right
-                (fun (branch, f, eff_label, cflags, c, g) (caccum, gaccum) ->
+                (fun (branch, f, eff_label, cflags, c, g, erasable_branch) (caccum, gaccum, erasable) ->
                      (f, eff_label, cflags, c)::caccum,
-                     Env.conj_guard g gaccum)
+                     Env.conj_guard g gaccum,
+                     erasable || erasable_branch)
                 t_eqns
-                ([], Env.trivial_guard) in
+                ([], Env.trivial_guard, false) in
       (* bind_cases adds an exhaustiveness check *)
-      TcUtil.bind_cases env res_t cases, g
+      TcUtil.bind_cases env res_t cases,
+      g,
+      erasable
     in
 
     let cres = TcUtil.bind e1.pos env (Some e1) c1 (Some guard_x, c_branches) in
+    let cres =
+      if erasable
+      then (* promote cres to ghost *)
+           let e = U.exp_true_bool in
+           let c = mk_GTotal' U.t_bool (Some U_zero) in
+           TcUtil.bind e.pos env (Some e) (U.lcomp_of_comp c) (None, cres)
```

Comment:
Nit: any reason to choose `bool` type here, and not `unit`? Also, `lift`ing `cres` to `GHOST` would have been another way to do this. If we do the `lift` here, then we can give a better error message (that it fails because the code is trying to pattern match on an erasable data constructor, currently we will get something like `effect GHOST and ST are not compatible` or something).

Reply:
Ah, for the first point because it will get promoted to `Tot unit` :).

Reply:
Regarding lift: how would you suggest doing it without inserting a bind? We could provide a new API call for lifting ... but there isn't one available at the moment.

Also, I'm not sure if we can really provide a better error message with a lift: you are allowed to match on an erasable type, incurring a ghost effect. But that ghost effect must be compatible with the effect of the branches and of the context.

I'm not sure how to attribute blame for a failed effect computation for a match+its context to, e.g., the scrutinee that was matched. But, whether we use a bind or a lift, we could attribute a failed effect computation for the scrutinee + its branches to the effect of matching the scrutinee. I can add something for that.